### PR TITLE
Create the accounts proof request

### PR DESCRIPTION
**Pull request by viquezclaudio**
Create the requests and handlers that are necessary to perform account proving.

Depends on #1462.

### CODE REVIEW

**Comment by viquezclaudio** [comment]
@sisou Just to let you know. This is failing to compile to WASM because there is a dependency of the accounts crate into the database.
So we need to remove that dependency as part of the accounts refactor in order to be able to use this functionality in the web client.

**Comment by sisou** [comment]
Would it be possible to structure the request message in a way that it allows reusing for querying stakers and validators? Right now it doesn't know the difference and I guess only returns the regular accounts. How can we query staker accounts and validator accounts?

**Comment by viquezclaudio** [comment]
LGTM

File: consensus/src/messages/handlers.rs [hunk 1]
```diff
@@ -476,3 +476,28 @@ impl Handle<ResponseTransactionReceiptsByAddress, Arc<RwLock<Blockchain>>>
...
+#[cfg(feature = "full")]
+impl Handle<ResponseAccountsProof, Arc<RwLock<Blockchain>>> for RequestAccountsProof {
+    fn handle(&self, blockchain: &Arc<RwLock<Blockchain>>) -> ResponseAccountsProof {
+        let blockchain = blockchain.read();
+
+        // We only prove accounts that exist in our current state
+        let proof = blockchain.get_accounts_proof(&self.addresses);
+
+        if proof.is_none() {
+            // If we could not generate a proof we response an empty result
```
Comment (sisou) [inline consensus/src/messages/handlers.rs]:
```suggestion
            // If we could not generate a proof we respond with an empty result
```

File: consensus/src/messages/handlers.rs [hunk 2]
```diff
@@ -476,3 +476,28 @@ impl Handle<ResponseTransactionReceiptsByAddress, Arc<RwLock<Blockchain>>>
...
+            return ResponseAccountsProof {
+                proof: None,
+                block_hash: None,
+            };
+        }
+
+        let block_hash = blockchain.head_hash();
```
Comment (sisou) [inline consensus/src/messages/handlers.rs:496]: Is this guaranteed to be the same hash that is used to generate the proof? No chance that the head is updated in this blockchain-read in the meantime?

File: consensus/src/messages/mod.rs [hunk 3]
```diff
@@ -248,3 +249,25 @@ pub struct ResponseTransactionReceiptsByAddress {
...
+    #[beserial(len_type(u16, limit = 128))]
+    /// Addresses for which the accounts trie proof is requested for
+    pub addresses: Vec<Address>,
+}
+
+impl RequestCommon for RequestAccountsProof {
+    type Kind = RequestMarker;
+    const TYPE_ID: u16 = 215;
+    type Response = ResponseAccountsProof;
+    const MAX_REQUESTS: u32 = MAX_REQUEST_TRANSACTIONS_PROOF;
```
Comment (sisou) [inline consensus/src/messages/mod.rs]: Are we sure we should reuse the TRANSACTIONS_PROOF limit for this request?

**Review by sisou — APPROVED** [review]

File: consensus/src/messages/handlers.rs (hunk 2 above)
Comment (jsdanielh) [reply consensus/src/messages/handlers.rs:496]: Yup, the read lock is taken while this happens

File: consensus/src/messages/mod.rs (hunk 3 above)
Comment (jsdanielh) [reply consensus/src/messages/mod.rs]: Value is correct but name is not

File: consensus/src/messages/mod.rs [hunk 4]
```diff
@@ -248,3 +251,25 @@ pub struct ResponseTransactionReceiptsByAddress {
     #[beserial(len_type(u16, limit = 128))]
     pub receipts: Vec<(Blake2bHash, u32)>,
 }
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct RequestTrieProof {
```
Comment (sisou) [inline consensus/src/messages/mod.rs:256]: The handler for this message still needs to be registered in `consensus/src/consensus/mod.rs`.

Comment (jsdanielh) [reply consensus/src/messages/mod.rs:256]: Thanks. I already had forgotten about it

**Review by jsdanielh — APPROVED** [review]

**Pull request merged**